DataViews: Add `between` operator

chihsuan opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Part of WOOA7S-123

## Proposed Changes

Add a new "between" filter operator for dataviews, supporting inclusive range filtering for numeric fields (e.g. price, item count, etc).

To implement this, this PR update the `Edit` to accept `operator` prop and update the Integer control to render Between UI when the operator is "between". The operator prop would be very useful for datetime field where we can a Calendar control for the between operator.

While this functionality is also compatible with datetime fields, the UI for date selection has not yet been implemented, so it is not enabled by default for date fields.


## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

Some use cases in Woo Analytics require filters that accept user-provided input — like "Order item count between 10 and 100".


## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Run `yarn dataviews:storybook:start`
* Go to http://localhost:57863/?path=/story/dataviews-dataviews--default
* Select the `Satellites` filter
* Select operator `between (inc)`


https://github.com/user-attachments/assets/923353c9-1b92-4b18-8dd6-09a05db0a60e

![image](https://github.com/user-attachments/assets/22527e53-37aa-4a3f-a566-a25dec29d189)


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages.
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

chihsuan committed:
Add BETWEEN operator for filtering and sorting

- Introduced OPERATOR_BETWEEN in constants and types.
- Implemented filtering logic for the BETWEEN operator in filter-and-sort-data-view.
- Updated input widget to handle BETWEEN operator with min and max values.
- Enhanced UI components to display BETWEEN filter correctly.
- Added tests to validate BETWEEN operator functionality for numbers and dates.

chihsuan committed:
Update readme

chihsuan committed:
Update OperatorSelector

chihsuan committed:
Use filter label

chihsuan committed:
Fix conflicts

chihsuan committed:
Fix tests

chihsuan commented on packages/dataviews/src/components/dataviews-filters/input-widget.tsx:
```diff
@@ -68,15 +66,73 @@ export default function InputWidget( {
+						<Notice status="warning" isDismissible={ false }>
+							{ __( 'Min value must be less than max value.' ) }
+						</Notice>
+					</FlexItem>
```
@Automattic/dotcom-design 👋  I'm working on a `between` operator that allows merchants to set a minimum and maximum value for filtering. Could I get your feedback on what to do if the minimum value exceeds the maximum value? At the moment, I'm considering displaying a warning message like this:


<img width="423" alt="Screenshot 2025-05-22 at 17 16 43" src="https://github.com/user-attachments/assets/7e79f55e-d738-4630-8a66-6a02b7d30ce5" />

jasmussen replied:
If the fields are always going to be small values, can you place them side by side? 

I'd also add labels, e.g. "Min." and "Max." and potentially help text or just an explanatory paragraph below. 

I also wonder if instead of a notice, you could use some of the validation patterns that @mirka has been working on. I.e. "invalid email" outputting something in red below a field; something similar might work here.

chihsuan replied:
Thanks for the quick response!  @jasmussen 

> If the fields are always going to be small values, can you place them side by side?

I think it's not always the case. It depends on the data. A large store might have large number of orders/revenue.

> I'd also add labels, e.g. "Min." and "Max." and potentially help text or just an explanatory paragraph below.

That's a good idea. 

> I also wonder if instead of a notice, you could use some of the validation patterns that @mirka has been working on. I.e. "invalid email" outputting something in red below a field; something similar might work here.

I'll try that. 🙌

mirka replied:
All this PR #100771 needs is a ✅ Approval from you @chihsuan (or @louwie17 👋) as the main consumer and we can merge the first iteration 😄 (Marco is swamped right now with higher priority stuff)

chihsuan replied:
I've approved PR #10071.

chihsuan replied:
I've added the labels and help text as suggested, and I set the `min` and `max` attributes for the input fields so the values automatically adjust based on user input. 

<img width="265" alt="Screenshot 2025-05-23 at 14 25 19" src="https://github.com/user-attachments/assets/e4dcdf0a-9590-424a-adb0-0c7b1fc6edb7" />

https://github.com/user-attachments/assets/c9b0b8ba-6cf4-49f6-812f-cfae7aee118f

jasmussen replied:
There's a tiny layout shift at 1s 2s in the video above (thank you for providing!) where it appears the title column grows a little wider. Do you know what that is?<br><br>Screenshots look good. Please:

* Add a period after Min → "Min."
* Add aperiod after Max → "Max."
* Add a period at the end of the help text.

So the help text would be:

> Max. value must be greater than min. value.

We might even want to massage it to this:

> The max. value must be greater than the min. value.

I still think we can have these two inputs in a single row, you can put a BIG number in such wide fields, and I don't think they need to be so wide. But not a blocker.

chihsuan replied:
> There's a tiny layout shift at 1s 2s in the video above (thank you for providing!) where it appears the title column grows a little wider. Do you know what that is?

It seems that when opening the filter dropdown, the page scroll is disabled.  🤔 

<img width="1049" alt="Screenshot 2025-05-26 at 13 17 56" src="https://github.com/user-attachments/assets/79855033-1b6c-4bfb-be12-4ded338e4f0c" />

chihsuan replied:
Thanks @jasmussen. I've updated the PR as you suggested. 🙂 

<img width="301" alt="Screenshot 2025-05-26 at 13 31 19" src="https://github.com/user-attachments/assets/12fa9e82-e127-49be-bc54-91a94bee12d3" />

jasmussen replied:
Looks good. Can the gap be $grid-unit-20? Otherwise, this just needs a quick code review.

chihsuan replied:
Sure. : ) 

Done in 20e66028a109b344fb03a849251952333ed79541

<img width="432" alt="Screenshot 2025-05-26 at 18 34 34" src="https://github.com/user-attachments/assets/9462774d-4034-4ab8-987d-59cbb3d5ddf6" />

chihsuan committed:
Add operator support for BETWEEN in Integer control

- Introduced BetweenControls component to handle min and max values.
- Updated Integer component to render BetweenControls when operator is BETWEEN.
- Enhanced InputWidget to pass the current filter operator.
- Added operator property to DataFormControlProps for better filter handling.

chihsuan committed:
Improve readability

chihsuan committed:
Remove unnecessary check

chihsuan committed:
Address design feedback

chihsuan committed:
Change gap to $grid-unit-20 (16px)

oandregal commented on packages/dataviews/src/utils.ts:
```diff
@@ -22,14 +23,24 @@ export function sanitizeOperators< Item >( field: NormalizedField< Item > ) {
 		ALL_OPERATORS.includes( operator )
 	);
 
+	// Don't allow between operator if the field has elements.
```
Why the between operator is incompatible with elements?

chihsuan replied:
At the moment, we support either presets (`elements`) or user input (`Edit`), but not both. And the current presets UI (multi-select) is not ideal for this.

<img width="327" alt="Screenshot 2025-05-27 at 10 09 15" src="https://github.com/user-attachments/assets/dda80419-7518-413f-ab98-ceed5af97ec8" />

oandregal commented on packages/dataviews/src/utils.ts:
```diff
@@ -22,14 +23,24 @@ export function sanitizeOperators< Item >( field: NormalizedField< Item > ) {
 		SINGLE_SELECTION_OPERATORS.includes( operator )
 	);
-	if ( hasSingleSelectionOperator ) {
+	if ( hasSingleSelectionOperator || operators.includes( OPERATOR_BETWEEN ) ) {
```
Why do we want to remove any multi-selection operator if the between operator is present?

chihsuan replied:
Hmm, you're right. The between operator can be applied when a multi-selection operator is present. Updated the logic in 355177da4ad2c6d8f2eb17bfd98ee5455dc64da5.

oandregal commented on packages/dataviews/src/filter-and-sort-data-view.ts:
```diff
@@ -217,6 +218,27 @@ export function filterSortAndPaginate< Item >(
 						);
 					} );
+				} else if (
+					filter.operator === OPERATOR_BETWEEN &&
```
Not necessary for this PR, just sharing thoughts to improve later.

I think there's an opportunity to formalize what an Operator is. For example, we could do:

```js
[ OPERATOR_IS ]: {
        key: 'is-filter',
        label: __( 'Is' ),
        filter: ( data, field, filter ) {}
},
```

and `filterSortAndPaginate` will delegate the filter to the proper operator, if there's a match.

Doing so would enable us to support 3rd party operators, should we ever need that. I also like that adding/removing/updating an operator would only require touching a single place.

chihsuan replied:
That's a good idea! I agree that formalizing what an Operator is would be beneficial.

oandregal commented on packages/dataviews/src/components/dataviews-filters/input-widget.tsx:
```diff
@@ -68,16 +65,21 @@ export default function InputWidget( {
 			<field.Edit
 				data={ data }
 				field={ field }
+				operator={ currentFilter.operator }
```
It feels off that we add support for filter ranges by coupling the Edit function to the current filter. What are alternatives to this that we can explore?

One small tweak would be adding a range property (false by default) to the Edit function. The Edit function would still absorb the complexity of rendering single or range controls, but the semantics are clearer.

chihsuan replied:
Originally, I was thinking of rendering two Edit functions for the between operator. However, I found it limiting the flexibility of UI. For example, here we want to render input type as number, but for date field, we'd want to render a calendar.

If we pass the operator into the Edit function, we can customize the UI based on the specific operator. This approach would also enable us to provide different controls for date fields as needed.

<img width="392" alt="Screenshot 2025-05-27 at 10 17 14" src="https://github.com/user-attachments/assets/b5c84006-6bee-4522-9341-ae9e485d7fce" />



<img width="299" alt="Screenshot 2025-05-27 at 10 17 19" src="https://github.com/user-attachments/assets/426c9300-1fbe-407a-bf32-a75e0a9bc33a" />

chihsuan replied:
That said, I share your concerns about how coupling the operator with the Edit function can make the semantics unclear. I think it's a trade-off, but I'm open to other suggestions.

oandregal replied:
> If we pass the operator into the Edit function, we can customize the UI based on the specific operator. This approach would also enable us to provide different controls for date fields as needed.

It seems useful to focus solely on support for integer in this PR, and address datetime separately (which may end using a different operator or none at all, depending on how the new calendar widget works with ranges).

So, that leaves with this issue: how do we change the `field.Edit` at runtime based on the operator?

chihsuan replied:
> So, that leaves with this issue: how do we change the field.Edit at runtime based on the operator?

This sounds like we could have a mapping from operator to field.Edit within the field type? 🤔

chihsuan replied:
@oandregal I've addressed the comments in the PR. Regarding the `field.Edit` <-> operator issue. 

I think an alternative solution could be introducing a mapping from operator to the Edit function in the field type.

```js
{
  Edit: 'integer', // Keep existing behavior
  filterBy: {
    operatorEdits: {
      [OPERATOR_BETWEEN]: IntegerRangeInput,
      // Could add more operator-specific components as needed
    }
  },
  enableSorting: true,
};
```

This raises a question: What if a user supplies a single-value Input component for the BETWEEN operator? That seems a bit odd. However, since the current implementation lets users provide any Edit component, it may not actually be an issue.

Any thoughts on this? This way, we could avoid coupling the `operator` to the `field.Edit` component.

oandregal replied:
I checked what happens when a field author defines a custom Edit function for a field of type integer:

```js
{
  type: 'integer'
  Edit: ( { data, field, onChange, hideLabelFromVision, operator } ) => {
    // custom input
  }
}
```

In this case, the field author has two choices:

- either support the default operators defined for the `integer` type (including `between`)
- restrict the operators supported by the field to a subset by using `filterBy.operators` (where `between` can be removed)

So, complex cases work.

This case also highlights that an alternative that doesn't pass the `operator` to the Edit function still needs the mapping to be public API — so that field authors can map `between` to the proper Edit function, which is a bit disconnected and has worse developer experience than the approach in this PR.

oandregal replied:
> What if a user supplies a single-value Input component for the BETWEEN operator?

This is also possible with the implementation on this PR. A field author uses the default operators (that includes `between`) but the Edit function it provides doesn't properly account for that. This is a bug in the field implementation that it's up to the field author to fix.

chihsuan committed:
Update operator sanitization logic

oandregal commented:
This is a nice operator to have, thanks for taking up this work.

I've finally caught up with the operator's work, and all the things we added have made them a bit fragile. It's not clear how operators relate to types. For example, what does it mean that a boolean type has a less than operator? I'd like to address that at https://github.com/Automattic/wp-calypso/pull/104040 and come back to this PR after. Does that sound good?

chihsuan committed:
Update src/normalize-fields.ts

oandregal commented on packages/dataviews/src/constants.ts:
```diff
@@ -105,6 +107,10 @@ export const OPERATORS: Record< Operator, { key: string; label: string } > = {
 	},
+	[ OPERATOR_BETWEEN ]: {
+		key: 'between-filter',
+		label: __( 'between (inc)' ),
```
```suggestion
		label: __( 'Between (inc)' ),
```

oandregal replied:
Otherwise, it'd be the only lowercased operator (see the storybook for integer field type at `yarn workspace @automattic/dataviews storybook:start`):

<img width="412" alt="Screenshot 2025-06-10 at 12 00 32" src="https://github.com/user-attachments/assets/df72f97b-390c-423f-9c1c-4845356ef1bc" />

oandregal commented on packages/dataviews/src/field-types/datetime.tsx:
```diff
@@ -40,7 +39,15 @@ export default {
+		validOperators: [
+			OPERATOR_IS,
+			OPERATOR_IS_NOT,
+			OPERATOR_BETWEEN,
```
How does the `between` operator works for datetime type fields (see in `yarn workspace @automattic/dataviews storybook:start`)?

https://github.com/user-attachments/assets/aaa36355-b9ef-4b85-89f6-6f9c5f203165

oandregal replied:
I just noticed this in the PR description:

> While this functionality is also compatible with datetime fields, the UI for date selection has not yet been implemented, so it is not enabled by default for date fields.

It's okay to only add support for integers. We shouldn't enable this operator until it works for datetime fields.

chihsuan replied:
Right, we should disable it for date fields until it works. Reverted changes in 1dbcf860c0a8cde5448f4e0c9cc23100387faa57.

oandregal commented on packages/dataviews/src/components/dataviews-filters/search-widget.tsx:
```diff
@@ -39,7 +39,8 @@ function normalizeSearchInput( input = '' ) {
 const EMPTY_ARRAY: [] = [];
-const getCurrentValue = (
+
+export const getCurrentValue = (
```
If we need to reuse it in both edit and input, perhaps move this to a `dataviews/filters/utils.js` and make search-widget and input-widget use it?

chihsuan replied:
Sounds good. I'll move this to a utils.js.

chihsuan replied:
Done in 24ba0aae3d7c8f38d39dbff13c0ad32f3e79586a.

oandregal commented on packages/dataviews/src/components/dataviews-filters/input-widget.tsx:
```diff
@@ -49,10 +50,6 @@ export default function InputWidget( {
 	const handleChange = useCallback(
 		( data: Record< string, any > ) => {
 			const nextValue = data[ field.id ];
-			if ( nextValue === currentValue ) {
```
Why do we need to remove this check? `currentValue` is not used in this file now, btw.

chihsuan replied:
Sorry, I was thinking of using `fastDeepEqual` for checking the equality instead, since the value could be any type. Updated in 3066468e9039c27d41243ad1e174e4e94f3b2a2b.

oandregal commented on packages/dataviews/src/normalize-fields.ts:
```diff
@@ -63,14 +64,24 @@ function getFilterBy< Item >(
 			validOperators.includes( operator )
 		);
 
+		// The `between` operator is not supported for preset layout.
```
what's the "preset layout"?

oandregal replied:
I suppose this means that the `between` operator is not compatible with having `elements` in a field. I've tested this by applying the following patch:

```diff
diff --git a/packages/dataviews/src/components/dataviews/stories/fixtures.tsx b/packages/dataviews/src/components/dataviews/stories/fixtures.tsx
index 859907cbf95..b44c6af57c0 100644
--- a/packages/dataviews/src/components/dataviews/stories/fixtures.tsx
+++ b/packages/dataviews/src/components/dataviews/stories/fixtures.tsx
@@ -706,6 +706,13 @@ export const fields: Field< SpaceObject >[] = [
                label: 'Satellites',
                id: 'satellites',
                type: 'integer',
+               elements: [
+                       { value: 1, label: '1' },
+                       { value: 2, label: '2' },
+                       { value: 3, label: '3' },
+                       { value: 4, label: '4' },
+                       { value: 5, label: '5' },
+               ],
                enableSorting: true,
        },
        {
```

and this is what I see (expected either the `between` operator not to be available or that using it will display the edit controls):


https://github.com/user-attachments/assets/f1c6871e-67ce-4ad7-ba74-9897c5f70399

chihsuan replied:
> I suppose this means that the between operator is not compatible with having elements in a field.

Yes, that’s correct.

> and this is what I see (expected either the between operator not to be available or that using it will display the edit controls):

Good catch!  I overlooked addressing the default field behavior while resolving the conflicts. It should not be available. Fixed in e18b77f769a96f4d2ef4cc6b5a636638aa8c2910.

chihsuan committed:
Update packages/dataviews/src/constants.ts

Co-authored-by: André <583546+oandregal@users.noreply.github.com>

chihsuan committed:
Revert datetime change

chihsuan committed:
Move getCurrentValue to dataviews/filters/utils.js

chihsuan committed:
Update filter logic to exclude 'between' operator when elements are provided; adjust default operators accordingly in normalize-fields.ts

chihsuan committed:
Revert changes

chihsuan committed:
Use fastDeepEqual to compare value

oandregal commented on packages/dataviews/src/normalize-fields.ts:
```diff
@@ -63,14 +64,24 @@ function getFilterBy< Item >(
 		if ( hasSingleSelectionOperator ) {
 			operators = operators.filter( ( operator ) =>
-				SINGLE_SELECTION_OPERATORS.includes( operator )
+				// The 'Between' operator is unique as it can be combined with single selection operators.
```
As a follow-up, I think we need to re-evaluate this logic that prevents us from having single and multi selection operators registered. I've created ARC-346 to track it.

oandregal reviewed (approved):
Thanks for your patience on this one, it is a bit of a gnarly issue that stress-tested the idea of using `Edit` for filtering.

I think there are two ideas that are related to this and that would be helpful to pursue:

- Filter by date ranges:

In this case, in addition to making the date calendar component work with both individual dates and ranges for different operators, we may also need the opposite. We may want users to just interact with the calendar widget and update the operators accordingly.

- Applying the same filter with different operators:

I think this would be an alternative implementation to this PR. For example, instead of having a `between` operator, users would apply `lessThanOrEqual` and `greaterThanOrEqual` operators simultaneously. This would be a more flexible and powerful approach because users could apply `lessThan` and `greaterThan` instead, which is not possible with this implementation — to support that we'd need to create a `betweenButNotEqual` operator, and the explosion of operators this PR may introduce concerns me a bit.

This alternate approach could take a while to figure out. We'd need to settle on the right interaction first. So, to unblock usage of DataViews in more places, I'm fine with us going forward with this direction — we could revisit later and add back-compatibility checks (migrate `between` to any newer system we have) if we ever change direction.

chihsuan committed:
Merge branch 'trunk' into add/dataviews-filter-between-operator

chihsuan committed:
Fix test

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/17567107

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Hi @chihsuan, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include all of the following strings:
- `<Name>%1$s between (inc): </Name><Value>%2$s-%3$s</Value>`
- `Between (inc)`
- `The max. value must be greater than the min. value.`
- `Min.`
- `Max.`

Thank you in advance!
